File: packages/guardian-rn/android/src/main/kotlin/com/guardian/rn/SessionKeyManager.kt
package com.guardian.rn

import android.security.keystore.KeyGenerator
import android.security.keystore.KeyProperties
import java.util.concurrent.atomic.AtomicBoolean
import javax.crypto.SecretKey

/**
 * Generates and holds the per-session HMAC-SHA256 key.
 * Android: backed by AndroidKeyStore when available (API 23+).
 * Per ADR-0003: getSessionKeyBytes() may be called only once per instance.
 */
class SessionKeyManager {

    private val keyDelivered = AtomicBoolean(false)
    private val keyAlias = "guardian_session_${System.nanoTime()}"
    private var rawKey: ByteArray? = null

    init {
        rawKey = generateKey()
    }

    /**
     * Returns the 32-byte session key.
     * @throws IllegalStateException if called more than once.
     */
    fun getSessionKeyBytes(): ByteArray {
        if (!keyDelivered.compareAndSet(false, true)) {
            throw IllegalStateException("GuardianError: session key already delivered — may only be called once per session")
        }
        val key = rawKey ?: throw IllegalStateException("GuardianError: session key not initialised")
        val copy = key.copyOf()
        key.fill(0)  // zero out the original after delivery
        rawKey = null
        return copy
    }

    private fun generateKey(): ByteArray {
        return try {
            val kg = KeyGenerator.getInstance(
                KeyProperties.KEY_ALGORITHM_HMAC_SHA256,
                "AndroidKeyStore"
            )
            kg.init(
                android.security.keystore.KeyGenParameterSpec.Builder(
                    keyAlias,
                    KeyProperties.PURPOSE_SIGN or KeyProperties.PURPOSE_VERIFY
                ).build()
            )
            val secretKey: SecretKey = kg.generateKey()
            // Export raw bytes for delivery to JS — only done once
            val ks = java.security.KeyStore.getInstance("AndroidKeyStore")
            ks.load(null)
            (ks.getKey(keyAlias, null) as? javax.crypto.SecretKey)?.encoded
                ?: fallbackKey()
        } catch (e: Exception) {
            // Fallback to in-memory SecureRandom if Keystore unavailable
            fallbackKey()
        }
    }

    private fun fallbackKey(): ByteArray {
        val key = ByteArray(32)
        java.security.SecureRandom().nextBytes(key)
        return key
    }
}

File: packages/guardian-rn/ios/Sources/GuardianRN/SessionKeyManager.swift
import Foundation
import Security

/// Generates and holds the per-session HMAC-SHA256 key (ADR-0003).
/// getSessionKeyBytes() may be called exactly once per instance.
final class SessionKeyManager {

    private var rawKey: [UInt8]?
    private var keyDelivered = false
    private let lock = NSLock()

    init() {
        rawKey = generateKey()
    }

    /// Returns the 32-byte session key.
    /// Throws if called more than once.
    func getSessionKeyBytes() throws -> [UInt8] {
        lock.lock()
        defer { lock.unlock() }

        guard !keyDelivered else {
            throw GuardianError.sessionKeyAlreadyDelivered
        }
        guard var key = rawKey else {
            throw GuardianError.sessionKeyNotInitialised
        }

        keyDelivered = true
        rawKey = nil

        let copy = key
        // Zero out the local buffer
        key.withUnsafeMutableBufferPointer { $0.initialize(repeating: 0) }
        return copy
    }

    private func generateKey() -> [UInt8] {
        var bytes = [UInt8](repeating: 0, count: 32)
        let status = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        guard status == errSecSuccess else {
            // Fallback: arc4random_buf (never used in production; logs a warning)
            arc4random_buf(&bytes, bytes.count)
            return bytes
        }
        return bytes
    }
}

enum GuardianError: Error {
    case sessionKeyAlreadyDelivered
    case sessionKeyNotInitialised
}
